#  Ejercicio 1
## Web Scraping para dos páginas web empleando BeutifulSoup y Selenium

### BeautifulSoup

#### 1. http://quotes.toscrape.com

In [1]:
# Solicitud del html a la página web de interés
import requests

url = 'http://quotes.toscrape.com'
page = requests.get(url)

print('Status Code 200: La solicitud ha sido Confirmada')
print('Status Code 404: No se ha encontrado la página web')
print('Status Code 500: Se ha producido algún error interno')
print()
print('Resultado de la solicitud:', page.status_code)


Status Code 200: La solicitud ha sido Confirmada
Status Code 404: No se ha encontrado la página web
Status Code 500: Se ha producido algún error interno

Resultado de la solicitud: 200


In [2]:
# Creación de objeto BeautifulSoup
from bs4 import BeautifulSoup

soup = BeautifulSoup(page.content)

# Visualización del código html de forma ordenada
print(soup.prettify())

<!DOCTYPE html>
<html lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   Quotes to Scrape
  </title>
  <link href="/static/bootstrap.min.css" rel="stylesheet"/>
  <link href="/static/main.css" rel="stylesheet"/>
 </head>
 <body>
  <div class="container">
   <div class="row header-box">
    <div class="col-md-8">
     <h1>
      <a href="/" style="text-decoration: none">
       Quotes to Scrape
      </a>
     </h1>
    </div>
    <div class="col-md-4">
     <p>
      <a href="/login">
       Login
      </a>
     </p>
    </div>
   </div>
   <div class="row">
    <div class="col-md-8">
     <div class="quote" itemscope="" itemtype="http://schema.org/CreativeWork">
      <span class="text" itemprop="text">
       “The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”
      </span>
      <span>
       by
       <small class="author" itemprop="author">
        Albert Einstein
       </small>
       <a href="/author/Albert

In [3]:
# Extracción de las frases
quotes = soup.find_all('div',class_='quote') # Para encontrar las las "class quote" en todas las "div"
citas = []

for quote in quotes:
    cita = quote.find('span', class_ ='text').text
    citas.append(cita)

citas

['“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”',
 '“It is our choices, Harry, that show what we truly are, far more than our abilities.”',
 '“There are only two ways to live your life. One is as though nothing is a miracle. The other is as though everything is a miracle.”',
 '“The person, be it gentleman or lady, who has not pleasure in a good novel, must be intolerably stupid.”',
 "“Imperfection is beauty, madness is genius and it's better to be absolutely ridiculous than absolutely boring.”",
 '“Try not to become a man of success. Rather become a man of value.”',
 '“It is better to be hated for what you are than to be loved for what you are not.”',
 "“I have not failed. I've just found 10,000 ways that won't work.”",
 "“A woman is like a tea bag; you never know how strong it is until it's in hot water.”",
 '“A day without sunshine is like, you know, night.”']

In [4]:
# Extracción de los autores
autores=[]

for quote in quotes:
    autor=quote.find('small', class_='author').text
    autores.append(autor)

autores

['Albert Einstein',
 'J.K. Rowling',
 'Albert Einstein',
 'Jane Austen',
 'Marilyn Monroe',
 'Albert Einstein',
 'André Gide',
 'Thomas A. Edison',
 'Eleanor Roosevelt',
 'Steve Martin']

In [5]:
#Extracción de los Tags
tags=[]
for quote in quotes:
    tag=quote.find('div',class_='tags').meta['content']
    tags.append(tag)
    
tags

['change,deep-thoughts,thinking,world',
 'abilities,choices',
 'inspirational,life,live,miracle,miracles',
 'aliteracy,books,classic,humor',
 'be-yourself,inspirational',
 'adulthood,success,value',
 'life,love',
 'edison,failure,inspirational,paraphrased',
 'misattributed-eleanor-roosevelt',
 'humor,obvious,simile']

In [6]:
# Creación de un dataframe
import pandas as pd
data = {'Cita': citas, 'Autor': autores, 'Etiquetas': tags}
frases_autores = pd.DataFrame(data)

frases_autores

Cita              Autor  \
0  “The world as we have created it is a process ...    Albert Einstein   
1  “It is our choices, Harry, that show what we t...       J.K. Rowling   
2  “There are only two ways to live your life. On...    Albert Einstein   
3  “The person, be it gentleman or lady, who has ...        Jane Austen   
4  “Imperfection is beauty, madness is genius and...     Marilyn Monroe   
5  “Try not to become a man of success. Rather be...    Albert Einstein   
6  “It is better to be hated for what you are tha...         André Gide   
7  “I have not failed. I've just found 10,000 way...   Thomas A. Edison   
8  “A woman is like a tea bag; you never know how...  Eleanor Roosevelt   
9  “A day without sunshine is like, you know, nig...       Steve Martin   

                                  Etiquetas  
0       change,deep-thoughts,thinking,world  
1                         abilities,choices  
2  inspirational,life,live,miracle,miracles  
3             aliteracy,books,classic,humor  
4                 be-yourself,inspirational  
5                   adulthood,success,value  
6                                 life,love  
7  edison,failure,inspirational,paraphrased  
8           misattributed-eleanor-roosevelt  
9                      humor,obvious,simile

#### 2. https://en.wikipedia.org/wiki/List_of_compositions_by_Dmitri_Shostakovich

In [7]:
# Solicitud del html a la página web de interés

page_2 = requests.get('https://en.wikipedia.org/wiki/List_of_compositions_by_Dmitri_Shostakovich')

print('Status Code 200: La solicitud ha sido Confirmada')
print('Status Code 404: No se ha encontrado la página web')
print('Status Code 500: Se ha producido algún error interno')
print()
print('Resultado de la solicitud:', page_2.status_code)


Status Code 200: La solicitud ha sido Confirmada
Status Code 404: No se ha encontrado la página web
Status Code 500: Se ha producido algún error interno

Resultado de la solicitud: 200


In [8]:
# Creación de objeto BeautifulSoup

soup_2 = BeautifulSoup(page_2.content)


# print(soup_2.prettify() ----> Para visualización del código html de forma ordenada (no ejecutado para evitar extensión)

In [10]:
# Búsqueda de la tabla que contiene la información de interés

tabla_wik = soup_2.find('table', class_ = 'wikitable')

In [11]:
# Selección de las columnas de interés con sus correspondientes registros o filas
Title = []
Instrumentation = []
Year = []

filas = tabla_wik.find_all("tr")[1:]  

for fila in filas:
    fils_tabl = fila.find_all("td")
    title = fils_tabl[1].text.strip()
    instruments = fils_tabl[2].text.strip()
    year = fils_tabl[3].text.strip()

    Title.append(title)
    Instrumentation.append(instruments)
    Year.append(year)

In [12]:
# Creación del dataframe con columnas de interés

data = {'Título': Title, 'Instrumentación': Instrumentation, 'Año': Year}
compositions_by_Shostakovich = pd.DataFrame(data)
compositions_by_Shostakovich



Título  \
0    Five Piano Pieces (The Soldier, Fiery Sonata, ...   
1                                      Hymn to Freedom   
2    Taras Bulba (based on the eponymous story by N...   
3                               Revolutionary Symphony   
4    Funeral March in Memory of the Victims of the ...   
..                                                 ...   
264         Suite on Verses of Michelangelo Buonarroti   
265                   Four Verses of Captain Lebyadkin   
266  Orchestration of Beethoven's "Es war einmal ei...   
267                         Sonata for Viola and Piano   
268                                    Symphony No. 16   

                     Instrumentación              Año  
0                              Piano        1914–1915  
1                              Piano        1915–1916  
2    Opera (instrumentation unknown)        1915–1916  
3                          Orchestra  1917–April 1918  
4                              Piano     January 1918  
..                               ...              ...  
264                   Bass and piano             1974  
265                   Bass and piano             1974  
266               Bass and orchestra             1975  
267                  Viola and piano             1975  
268                        Orchestra             1975  

[269 rows x 3 columns]

Se ha realizado una extracción de la información relativa a las composiciones de Dmitri Shostakovich, uno de los grandes compositores del S. XX. La tabla contiene ahora solo el "título", "instrumentación" y el "año".

### Selenium

#### 1. http://quotes.toscrape.com

In [13]:
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys

# Iniciación del controlador y acceso a la url
driver = webdriver.Chrome()
driver.get('http://quotes.toscrape.com/')

In [14]:
# Extracción de las frases
citas_slnium = driver.find_elements(By.CLASS_NAME, 'text')

for cita in citas_slnium:
    cita = cita.text
    print(cita)

“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”
“It is our choices, Harry, that show what we truly are, far more than our abilities.”
“There are only two ways to live your life. One is as though nothing is a miracle. The other is as though everything is a miracle.”
“The person, be it gentleman or lady, who has not pleasure in a good novel, must be intolerably stupid.”
“Imperfection is beauty, madness is genius and it's better to be absolutely ridiculous than absolutely boring.”
“Try not to become a man of success. Rather become a man of value.”
“It is better to be hated for what you are than to be loved for what you are not.”
“I have not failed. I've just found 10,000 ways that won't work.”
“A woman is like a tea bag; you never know how strong it is until it's in hot water.”
“A day without sunshine is like, you know, night.”


In [15]:
# Extracción de los autores
autores_slnium = driver.find_elements(By.CLASS_NAME, 'author')

for autor in autores_slnium:
    autor= autor.text
    print(autor)

Albert Einstein
J.K. Rowling
Albert Einstein
Jane Austen
Marilyn Monroe
Albert Einstein
André Gide
Thomas A. Edison
Eleanor Roosevelt
Steve Martin


In [16]:
# Extracción de los tags

tags_slnium = driver.find_elements(By.CLASS_NAME, "tags")

for tag in tags_slnium:
    tag= tag.text
    print(tag)

Tags: change deep-thoughts thinking world
Tags: abilities choices
Tags: inspirational life live miracle miracles
Tags: aliteracy books classic humor
Tags: be-yourself inspirational
Tags: adulthood success value
Tags: life love
Tags: edison failure inspirational paraphrased
Tags: misattributed-eleanor-roosevelt
Tags: humor obvious simile


In [17]:
# Creación de un dataframe
data_slnium = {'Cita': [cita.text for cita in citas_slnium],
               'Autor': [autor.text for autor in autores_slnium],
               'Tags': [tag.text for tag in tags_slnium]} 

frases_autores_slnium = pd.DataFrame(data_slnium) 
frases_autores_slnium['Tags'] = frases_autores_slnium['Tags'].str.replace('Tags: ', '')

frases_autores_slnium     


Cita              Autor  \
0  “The world as we have created it is a process ...    Albert Einstein   
1  “It is our choices, Harry, that show what we t...       J.K. Rowling   
2  “There are only two ways to live your life. On...    Albert Einstein   
3  “The person, be it gentleman or lady, who has ...        Jane Austen   
4  “Imperfection is beauty, madness is genius and...     Marilyn Monroe   
5  “Try not to become a man of success. Rather be...    Albert Einstein   
6  “It is better to be hated for what you are tha...         André Gide   
7  “I have not failed. I've just found 10,000 way...   Thomas A. Edison   
8  “A woman is like a tea bag; you never know how...  Eleanor Roosevelt   
9  “A day without sunshine is like, you know, nig...       Steve Martin   

                                       Tags  
0       change deep-thoughts thinking world  
1                         abilities choices  
2  inspirational life live miracle miracles  
3             aliteracy books classic humor  
4                 be-yourself inspirational  
5                   adulthood success value  
6                                 life love  
7  edison failure inspirational paraphrased  
8           misattributed-eleanor-roosevelt  
9                      humor obvious simile

In [18]:
# Cierre del controlador de Selenium
driver.quit()

#### 2. https://en.wikipedia.org/wiki/List_of_compositions_by_Dmitri_Shostakovich

In [19]:
# Iniciación del controlador y acceso a la url
driver_2 = webdriver.Chrome()
driver_2.get('https://en.wikipedia.org/wiki/List_of_compositions_by_Dmitri_Shostakovich')

In [20]:
# Búsqueda de la tabla que contiene la información de interés

tabla_wik_slnium = driver_2.find_element(By.CSS_SELECTOR, '.wikitable')

In [21]:
# Selección de las columnas de interés con sus correspondientes registros o filas
Title = []
Instrumentation = []
Year = []

filas_slnium = tabla_wik_slnium.find_elements(By.TAG_NAME, "tr")[1:]

for fila in filas_slnium:
    fils_tabl_slnium = fila.find_elements(By.TAG_NAME, "td")
    title = fils_tabl_slnium[1].text.strip()
    instruments = fils_tabl_slnium[2].text.strip()
    year = fils_tabl_slnium[3].text.strip()

    Title.append(title)
    Instrumentation.append(instruments)
    Year.append(year)

In [22]:
# Creación del dataframe con columnas de interés

data_slnium = {'Título': Title, 'Instrumentación': Instrumentation, 'Año': Year}
compositions_by_Shostakovich = pd.DataFrame(data_slnium)
compositions_by_Shostakovich



Título  \
0    Five Piano Pieces (The Soldier, Fiery Sonata, ...   
1                                      Hymn to Freedom   
2    Taras Bulba (based on the eponymous story by N...   
3                               Revolutionary Symphony   
4    Funeral March in Memory of the Victims of the ...   
..                                                 ...   
264         Suite on Verses of Michelangelo Buonarroti   
265                   Four Verses of Captain Lebyadkin   
266  Orchestration of Beethoven's "Es war einmal ei...   
267                         Sonata for Viola and Piano   
268                                    Symphony No. 16   

                     Instrumentación              Año  
0                              Piano        1914–1915  
1                              Piano        1915–1916  
2    Opera (instrumentation unknown)        1915–1916  
3                          Orchestra  1917–April 1918  
4                              Piano     January 1918  
..                               ...              ...  
264                   Bass and piano             1974  
265                   Bass and piano             1974  
266               Bass and orchestra             1975  
267                  Viola and piano             1975  
268                        Orchestra             1975  

[269 rows x 3 columns]

In [23]:
# Cierre del controlador de Selenium
driver_2.quit()

# Ejercicio 2
## Documentar en un Word el conjunto de datos generado con la información que tienen los diferentes archivos de Kaggle (adjunto repositorio)

## Sobre Dataset "Quotes to Scrape"
### Contexto

El dataset contiene citas célebres de diferentes personajes historicos, el nombre de estos y varias etiquetas que guardan relación con la correspondiente cita. La información contenida pertenece a una página web creada con fines educativos para la práctica del web scraping.


### Contenido

10 filas y 3 columnas ("Cita", "Autor" y "Etiquetas")


### Fecha de extracción de la información

27 de septiembre de 2023

#### -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

## Sobre Dataset "List of compositions by Dimitri Shostakovich"
### Contexto

El dataset contiene todas las obras pertenecientes a Dmitri Shostakovich, compositor de la entonces Unión Soviética y uno de los más importantes de S. XX. Su producción abarca todos los géneros, desde la ópera, sinfonías, conciertos para instrumentos, cantatas, cuartetos de cuerda e incluso pasando por la música para cine.
El dataset se estructura por el "Título" de cada obra, la "Instrumentación" empleada y el "Año" de creación. La información se ha obtenido de la "wikipedia", una enciclopedia online abierta a todo el mundo y generada de manera colaborativa.


### Contenido

269 filas y 3 columnas ("Título", "Instrumentación" y "Año")


### Fecha de extracción de la información

27 de septiembre de 2023

# Ejercicio 3
## Elección de una página web y realización de web scraping con la librerería Selenium  O en Scrapy


### Selenium para extraer tabla de cotización de divisas en https://es.investing.com/currencies/exchange-rates-table

In [24]:
# Iniciación del controlador y acceso a la url
driver_divisas = webdriver.Chrome()
driver_divisas.get('https://es.investing.com/currencies/exchange-rates-table')

In [25]:
# tabla_divisas = driver_divisas.find_element(By.CSS_SELECTOR, '.divisas_europa')

In [26]:
# Búsqueda de la tabla que contiene la información de interés

tabla_divisas = driver_divisas.find_element(By.CSS_SELECTOR,  '.genTbl')


In [27]:
# Selección de las columnas de interés con sus correspondientes registros o filas
Par = []
DOLAR= []
EURO = []
LIBRA = []

filas_divisas = tabla_divisas.find_elements(By.TAG_NAME, "tr")[1:]

for fila in filas_divisas:
    fils_tabl_divisas = fila.find_elements(By.TAG_NAME, "td")
    par = fils_tabl_divisas[0].text.strip()
    dolar = fils_tabl_divisas[1].text.strip()
    euro = fils_tabl_divisas[2].text.strip()
    libra = fils_tabl_divisas[3].text.strip()

    Par.append(par)
    DOLAR.append(dolar)
    EURO.append(euro)
    LIBRA.append(libra)

In [28]:
# Creación de un dataframe
data_divisas = {'Euro': EURO,
               'Dólar': DOLAR,
               'Libra': LIBRA} 

divisas = pd.DataFrame(data_divisas, index=Par) 

divisas


Euro   Dólar     Libra
USD  0,9461       1    0,8228
EUR       1  1,0569    0,8697
GBP  1,1499  1,2153         1
JPY  0,0063  0,0067  0,552150
CHF  1,0331  1,0918    0,8983
CAD  0,6998  0,7396    0,6086
AUD  0,6038  0,6381    0,5252
NZD  0,5617  0,5937    0,4886

In [29]:
# Cierre del controlador de Selenium
driver_divisas.quit()

La tabla muestra las cotizaciones de las divisas del Euro, Dolar y la Libra esterlina con respecto a otras divisas a fecha del 27 de septiembre del 2023 y hora 10:15, ya que estamos ante la extracción de información de una tabla dinámica.